specs/HostResourceMapping.md

```diff
@@ -0,0 +1,108 @@
+# Background
+We have received feedbacks that there is a need to access local files in WebView2. In this document we describe the APIs for this scenario. We'd appreciate your feedback.
+
+
+# Description
+To access local files in WebView2, define a virutal host name to local folder mapping using the `AddVirtualHostNameToFolderMapping` API,
```

Comment:
Why use a virtual hostname rather than a scheme name?   The file: scheme is faulty as you well know (since you wrote the doc on that :), but I don't see why another scheme name cannot be used.    That would avoid overloading the hostname field, and also get rid of the need to maintain a mapping table.

Reply:
We had a separate conversation in mail which I'll try to summarize here that we can add to the background section of this doc.

In EdgeHtml we already had ms-appx mapping to app install location, ms-appdata with multiple hostname options to map to different kinds of app data paths, and requests for additional URI scheme to path mappings. We can provide the more flexible option to the end developer to set their own mappings to avoid us adding several new URI schemes and still not satisfying everyone.

And now we support non-UWP apps and additional install technologies which means we need to allow the end developer to set the folder paths themselves since we won't always know the correct paths for app data or app install paths.

Different paths can contain files with varying trust and privacy concerns. In UWP package app install locations contained files that we knew matched what had been validated by the Store so were trusted and had no privacy concerns in that the Store would give the same package to everyone. But the app data folder had no such guarantees. Now supporting non-UWP we must allow the end developer to tell us CORS policy and we must allow them to give different paths different origins in order to properly separate or not separate content as appropriate.

We had originally considered using a new URI scheme (although still with an API to set the hostname to folder mapping and CORS policy) but ultimately decided that using virtual hostnames for http(s) URIs was a better choice because:
1. Browsers have many checks scattered throughout their code for http and https URIs. Browsers require https for access to a small and growing list of APIs including geolocation, getUserMedia, web crypto, and payment APIs. Browsers suppress the referer HTTP header for privacy reasons when going from https to http. Using virtual hostnames instead of a new URI scheme means end developers don't have to question how it will work since they're just using http(s).
2. On a similar note, the implementation will be much more reliable since we won't have to seek out every location where a check is made for http or https.
3. This works similarly to the Android WebView's [WebViewAssetLoader](https://developer.android.com/reference/androidx/webkit/WebViewAssetLoader).